src/qibo/parameter.py

```diff
@@ -0,0 +1,109 @@
+import numpy as np
+import sympy as sp
+
+from qibo.config import raise_error
+
+
+class Parameter:
+    """Object which allows for variational gate parameters. Several trainable parameter
+    and possibly features are linked through a lambda function which returns the
+    final gate parameter. All possible analytical derivatives of the lambda function are
+    calculated at the object initialisation using Sympy.
+
+    Example::
+
+        from qibo.parameter import Parameter
+        param = Parameter(
+                lambda x, th1, th2, th3: x**2 * th1 + th2 * th3**2,
+                [1.5, 2.0, 3.0],
+                feature=[7.0],
+            )
+
+        partial_derivative = param.get_partial_derivative(3)
+
+        param.update_parameters(trainable=[15.0, 10.0, 7.0], feature=[5.0])
+        gate_value = param()
+
+
+    Args:
+        func (function): lambda function which builds the gate parameter. If both features and trainable parameters
+        compose the function, it must be passed by first providing the features and then the parameters, as
```

Comment:
1. I think it is still a bit ambiguous, because we are asking to write the `lambda` with features followed by parameters, but then we call the `Parameter(function, trainable, feature)`. I think is maybe better to keep the order `features` $\to$ `trainable` also in the initialisation. 
2. At this point I would call the features vector with the plural form `self.features`, because we can need to inject more than one feature.
3. I was thinking it can maybe be useful to set some default initialization of the parameters, e.g. to initialize them to 0 or to a random value if none is provided by the user.

Reply:
I can implement 1. and 2. without any problems. However, for 3. since Ale preferred not to do any checks on the number of features and the number of parameters, I can't initialise the parameters, since I do not have any information on the size of the parameter list

Reply:
Ok, but on this way the initialization is accepted only if the user provides the values too. Is this what you mean @WanderingMike, @AleCandido?

Reply:
Sorry @WanderingMike, I didn't mean to be so strict.
I always try to provide the rationale for my suggestions, but I might have failed. In any case, you do not need to do anything just because I said something :)

In any case, about checking, I would still avoid. It will fail anyhow, when it tries to use it, with less code on our side. But nothing prevents you to collect the same information for the suggested initialization. This is something that Python won't do for us, if we do not instruct it to do it.

About the initialization itself, I'm not sure whether it is worth it to do it silently. Maybe even requiring it from the user is not a completely bad idea.

Reply:
> About the initialization itself, I'm not sure whether it is worth it to do it silently. Maybe even requiring it from the user is not a completely bad idea.

In case we don't allow to do it silently, I would suggest to raise a custom error, asking explicitly to provide values for both trainable parameters and features. 

Despite this, I still believe the sense of `Parameter` is more in its being flexible (a lambda) than in the value it returns. On top of this idea, I think it should be possible to initialize an "empy" `Parameter`, which e.g. symbolically returns zero, but still exists. I am thinking to a QML scenario in which I need many parameters but I don't really care about the initial values.

Reply:
I implemented an initialisation with an array full of 0.0 float values